File: Data Structures/BinaryTree/C/binary_tree.c
#include "binary_tree.h"

/*
    This method handles inserting a new node in the binary tree
    Pass a pointer to the root of the tree
*/
struct binaryTree *insert(struct binaryTree *root, int data) {
    // if root is null, we have no other nodes
    if (root == NULL) {
        // allocate memory
        root = (struct binaryTree *) malloc((sizeof (struct binaryTree)));

        // if root is null
        if (root == NULL) {
            // we have an error, return
            fprintf(stderr, "Memory Error constructing the root of the tree\n");
            return NULL;
        } else {
            // set the data and left and right to null
            root->data = data;
            root->left = root->right = NULL;
        }
    } else {
        // if the data is less than the root
        if (data < root->data) {
            // recursive call to insert on the left side
            root->left = insert(root->left, data);
        // if the data is greater than the root
        } else if (data > root->data) {
            // recursive call to insert on the right side
            root->right = insert(root->right, data);
        }
    }
    // return the root
    return root;
}

/*
    This method finds the data in quesiton in the tree
*/
int find(struct binaryTree *root, int data) {
    // if root is empty
    if (root == NULL){
        // output message return false
        printf("Tree is empty\n");
        return 0;
    }

    // while root isn't null
    while (root) {
        // if data is the current data
        if (data == root->data)
            // return true
            return 1;
        // else if the data is greater
        else if (data > root->data)
            // set root to next on the right
            root = root->right;
        // else if the data is less
        else
            // set root to next on the right
            root = root->left;
    }
    // fall out of while loop, return false
    return 0;
}

/*
    This meghot finds the data in question within the tree
    uses recursion to find the data
*/
struct binaryTree *findRecursive(struct binaryTree *root, int data) {
    // base case:
    // if root is null
    if (root == NULL) {
        // output message and return null
        printf("Tree is empty\n");
        return NULL;
    }

    // If data is less than the roots data
    if (data < root->data)
        // recurse to the left
        return findRecursive(root->left, data);
    // if data is greater
    else if (data > root->data)
        // recurse on the right
        return findRecursive(root->right, data);
    // finally return the root
    return root;
}

/*
    This method outputs the tree in pre order traversal
    uses recursion to accomplish
*/
void preOrderTraversal(struct binaryTree *root) {
    if (root) {
        printf("%i ", root->data);
        preOrderTraversal(root->left);
        preOrderTraversal(root->right);
    }
}

/*
    This method outputs the tree in order traversal
    uses recursion
*/
void inOrderTraversal(struct binaryTree *root) {
    if (root) {
        inOrderTraversal(root->left);
        printf("%i ", root->data);
        inOrderTraversal(root->right);
    }
}

/*
    This method outputs the tree in post order traversal
    uses recursion
*/
void postOrderTraversal(struct binaryTree *root) {
    if (root) {
        postOrderTraversal(root->left);
        postOrderTraversal(root->right);
        printf("%i ", root->data);
    }
}

/*
    This method finds the max within the tree
*/
int findMax(struct binaryTree *root) {
    // TODO: change this to INT_MIN ?
    int rootVal, left, right, max = -99999;

    // if the root is not null
    if (root != NULL) {
        // set the temp variables
        rootVal = root->data;
        // recursively:
        // find max on left
        left = findMax(root->left);
        // find max on right
        right = findMax(root->right);

        // if left is greater than right, set max
        if (left > right)
            max = left;
        // if right is greater, set max
        else 
            max = right;
        // if the root is the greatest, set max
        if (rootVal > max)
            max = rootVal;
    }
    // return the max
    // TODO: if was root was null (?) it will return INT_MIN
    return max;
}

/*
    This method returns the size of the tree
*/
int size(struct binaryTree *root) {
    // if root is null, tree is emtpy
    if (root == NULL)
        return 0;
    else
        // we add one here for current node
        return size(root->left) + 1 + size(root->right);
}

/*
    This method returns the height of the tree
    Uses recursion
*/
int treeHeight(struct binaryTree *root) {
    int left, right;
    // if root is null, 0 size
    if (root == NULL)
        return 0;
    else {
        // recurse on the left side
        left = treeHeight(root->left);
        // recurse on the right side
        right = treeHeight(root->right);
        // if left is greater return left
        if (left > right)
            return left + 1;
        // right is greater, return
        else
            return right + 1;
        // +1 to the above because the root
    }
}

File: Data Structures/BinaryTree/Java/BinaryTree.java
public class BinaryTree<T extends Comparable<T>> {

    private int N;          // count of items in tree
    private Node<T> root;   // the root of the tree

    /**
     * The constructor
     *  initialize the private variables
     */
    public BinaryTree() {
        N = 0;
        root = null;
    }

    /**
     * This method inserts a new node
     * into the tree
     * @param item, the value to input into the tree
     */
    public void insert(T item) {
        // call private method to insert node
        root = insert(root, item);
        // increment the count of items in tree
        N++;
    }

    /**
     * This is the private method to handle inserting
     * a new node into the tree
     * @param node the current node (recursive)
     * @param item the value to input into tree
     * @return the node
     */
    private Node<T> insert(Node<T> node, T item) {
        // if current node is null
        if (node == null) {
            // return the new node with value
            return newNode(item);
        }

        // if item is less than current data
        if (item.compareTo(node.data) < 0)
            // insert to the left
            node.left = insert(node.left, item);
        // else the item is greater
        else if (item.compareTo(node.data) > 0)
            // insert to the right
            node.right = insert(node.right, item);
        // finally return the node
        return node;
    }

    /**
     * Private helper method to create
     * a new node
     * @param item value for new node
     * @return the new node
     */
    private Node<T> newNode(T item) {
        // initialize a new node
        Node<T> node = new Node<T>();
        // set the value
        node.data =item;
        // set left and right to null
        node.left = node.right = null;
        // return the new node
        return node;
    }

    /**
     * @return the size of nodes in the tree
     */
    public int size() { return N; }


    /**
     * Public interface to the height of the tree
     * @return the height of the tree
     */
    public int height() { return height(root);}

    /**
     * Private method to hanlde calculating
     * the height of the tree
     * @param node current node (recursive)
     * @return the height of the tree
     */
    private int height(Node<T> node) {
        // if the node is null, return 0
        if (node == null) return 0;
        else
            // get the max of the recursive calls of left and right side
            // we are getting the height of the tree, the most is the height
            // add one for the root
            return 1 + Math.max(height(node.left), height(node.right));
    }

    /**
     * Public interface to find the item
     * in the tree
     * @param item trying to find
     * @return true if in tree
     */
    public boolean find(T item) {
        return find(root, item);
    }

    /**
     * Private method to find the item in tree
     * @param node root node
     * @param item item searching for
     * @return true if found
     */
    private boolean find(Node<T> node, T item) {
        // if node is null, not in tree
        if (node == null) return false;
        
        // traverse tree as long as not null
        while (node != null) {
            // if we found, return true
            if (item == node.data) return true;
            // if less than, go left
            else if (item.compareTo(node.data) < 0)
                node = node.left;
            // otherwise, greater and go right
            else
                node = node.right;
        }
        // fall out of while loop, not in tree return false
        return false;
    }

    /**
     * Public method to traverse the tree in order
     */
    public void inOrderTraversal() {
        // call private method on root
        inOrderTraversal(root);
        // output new line
        System.out.println();
    }

    /**
     * Private method to traverse the tree in order
     * @param node current node
     */
    private void inOrderTraversal(Node<T> node) {
        // if node not null
        if (node != null) {
            // traverse to the left
            inOrderTraversal(node.left);
            // output the data and a space
            System.out.print(node.data + " ");
            // traverse to the right
            inOrderTraversal(node.right);
        }
    }

    /**
     * Public method for post order traversal
     */
    public void postOrderTraversal() {
        // call private method on root
        postOrderTraversal(root);
        // output new line
        System.out.println();
    }

    /**
     * Private method for post order traversal
     * @param node current node
     */
    private void postOrderTraversal(Node<T> node) {
        // if node isn't null
        if (node != null) {
            // traverse left and right, then output data
            postOrderTraversal(node.left);
            postOrderTraversal(node.right);
            System.out.print(node.data + " ");
        }
    }

    /**
     * Public method for pre order traversal
     */
    public void preOrderTraversal() {
        // call private method on the root
        preOrderTraversal(root);
        // output new line
        System.out.println();
    }

    /**
     * Private method for pre order traversal
     * @param node current node
     */
    private void preOrderTraversal(Node<T> node) {
        // if node isn't null
        if (node != null) {
            // output the data first, then traverse left and right
            System.out.print(node.data + " ");
            preOrderTraversal(node.left);
            preOrderTraversal(node.right);
        }
    }


    /**
     * Helper class Node with generic type
     * @param <T>
     */
    private class Node<T extends Comparable<T>> implements Comparable<T> {
        
        T data;         // the data
        Node<T> left;   // left node pointer
        Node<T> right;  // right node pointer

        /**
         * Helper method to compare the param
         * and current data
         * @param t data comparing to current
         * @return -1, 0, 1 depending on equality
         */
        @Override
        public int compareTo(T t) {
            return this.data.compareTo(t);
        }
    }
}

